### PR TITLE
CASTemplate as a generic template for CAS related operations

### PR DESCRIPTION
**Pull request by AmitKumarDas**
CASTemplate as a generic template for CAS related operations

- CAS template continues to be a feature gated in this commit
- CAS template exposes rest APIs for GET, LIST & DELETE of CAS volume
- New custom errors are introduced:
  - template.NotFoundError
  - template.VerifyError
- Above template based custom errors are set during cas template execution
  - Logic to set these errors are specified in the RunTask templates
- New top level properties introduced for template engine
  - JsonResult - To store the output of RunTask execution in []byte format
  - ListItems - To store a list of items e.g. list of output of various RunTasks
- New sub level properties w.r.t TaskResult top level property introduced
  - verifyErr - to store verification error message during template execution
  - notFoundErr - to store not found error message during template execution
- New sub level property w.r.t ListItems top level property introduced
  - currentRepeatResource - To store resource that is currenly being executed. 
This is set when a task is configured to be executed in a repetitive manner.
- A few labels have been deprecated for new labels corresponding to each of these deprecated labels
  - k8s.io/namespace to openebs.io/namespace
  - k8s.io/pvc to openebs.io/pvc
  - k8s.io/storage-class to openebs.io/storage-class
- Updates to CASTemplate spec & RunTask spec
  - taskNamespace & output have been re-positioned in the cas template spec
  - tasks are now only an array of task names in the cas template spec
- Introduces RunTask struct in pkg/task namespace
  - Earlier logic used a map[string]string to represent a RunTask
- Introduces 'post' property in RunTask
  - This is a set of template functions that are executed post execution of the RunTask
  - This deprecates the 'queries' property found in MetaTask specifications
- Introduces 'id' property in MetaTask specs
  - A RunTask will define its own 'id' property
  - This removes the need of defining a id for a task in CASTemplate
- Introduces 'repeatWith' property in MetaTask specs
  - RepeatWith sets one or more resources of the same kind that leads to repetitive execution of a task.
  - In other words a task template is executed multiple times.
  - The repetition count is determined by the resource names specified.
  - Each task execution is based on a resource from this provided list of resources.
- A bunch of go template functions are introduced
  - The intention is to make RunTask templating easier
  - isLen - flags if lenth of an object matches the given value
  - jsonPath - parses a json doc for specified properties
  - noop - swallows any input and returns an empty string
  - pickSuffix - picks the first item from a given list of items based on string suffix match
  - pickPrefix - picks the first item from a given list of items based on string prefix match
  - pickContains - picks the first item from a given list of items based on string contains match
  - asKeyMap - builds a map of kv pairs corresponding to each item of a given list of items
  - asNestedMap - builds a nested map corresponding to each item of a given list of items
  - saveAs - stores the given value at specified path of a destination object
  - notFoundErr - returns error only if the given object is empty
  - verifyErr - returns error if verification failed flag is true
  - toYaml - returns a YAML doc corresponding to the given interface{} object

Signed-off-by: AmitKumarDas <amit.das@openebs.io>

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/template/template.go [hunk 1]
```diff
@@ -19,37 +20,621 @@ package template
...
+//
+// Assumption here is .Values is of type 'map[string]interface{}'
+func verifyErr(errMessage string, hasVerificationFailed bool) (err error) {
+	if !hasVerificationFailed {
+		// no error if verification did pass successfully
+		return
+	}
+
+	if len(errMessage) == 0 {
+		errMessage = "verification failed"
```
Comment (gkGaneshR) [inline pkg/template/template.go]: if possible we can have the list of error messages in a common location or with typed string.

File: pkg/template/template.go [hunk 2]
```diff
@@ -19,37 +20,621 @@ package template
...
 //  https://github.com/kubernetes/helm/blob/master/pkg/chartutil/files.go
-func toYaml(yml string) string {
-	var m interface{}
-	err := yaml.Unmarshal([]byte(yml), &m)
+func toYaml(v interface{}) (yamlstr string) {
+	data, err := yaml.Marshal(v)
 	if err != nil {
-		// Swallow errors inside of a template.
-		return ""
+		// error is swallowed
```
Comment (gkGaneshR) [inline pkg/template/template.go]: can u pls explain why the error is swallowed? If we return the error, we can get some info at which line this yaml marshal has failed.

File: cmd/maya-apiserver/app/server/volume_endpoint.go [hunk 3]
```diff
@@ -21,18 +21,15 @@ const (
 // VolumeSpecificRequest is a http handler implementation. It deals with HTTP
 // requests w.r.t a single Volume.
 func (s *HTTPServer) volumeSpecificRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	// check the feature gate & switch if enabled
+	if util.CASTemplateFeatureGate() {
+		return s.volumeV1alpha1SpecificRequest(resp, req)
+	}
+
 	glog.Infof("received volume request: method '%s'", req.Method)
```
Comment (avishnu) [inline cmd/maya-apiserver/app/server/volume_endpoint.go]: The above message will no longer be logged. If there is a log in s.volumeV1alpha1SpecificRequest(resp, req), then this comment can be ignored.

File: cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go [hunk 4]
```diff
@@ -228,10 +235,10 @@ func (v *volumeAPIOpsV1alpha1) list() (*v1alpha1.CASVolumeList, error) {
 
 	cvols, err := vOps.List()
 	if err != nil {
-		glog.Errorf("failed to list cas template based volumes: error '%s'", err.Error())
+		glog.Errorf("failed to list cas template based volumes at namespaces '%s': error '%s'", vols.Namespace, err.Error())
 		return nil, CodedError(500, err.Error())
 	}
 
-	glog.Infof("cas template based volumes were listed successfully")
+	glog.Infof("cas template based volumes were listed successfully: namespaces '%s'", vols.Namespace)
```
Comment (avishnu) [inline cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go]: Will the value of vols.Namespace be populated in the following scenarios:
1) ALL namespace.
2) DEFAULT namespace.

File: pkg/apis/openebs.io/v1alpha1/cas_template_keys.go [hunk 5]
```diff
@@ -31,39 +31,54 @@ package v1alpha1
...
-	// TaskResultTLP is a top level property supported by
-	// volume policy engine
+	// The task related properties as placed with TaskTLP as the top level
+	// property
+	//TaskTLP TopLevelProperty = "Task"
+	// TaskResultTLP is a top level property supported by CAS template engine
 	//
-	// The specific results after the execution of a task
-	// as placed with TaskResultTLP as the top level property
+	// The specific results after the execution of a task as placed with
```
Comment (avishnu) [inline pkg/apis/openebs.io/v1alpha1/cas_template_keys.go]: Are the wordings right? "are placed with" instead of "as placed with" ?

File: pkg/apis/openebs.io/v1alpha1/cas_template_keys.go [hunk 6]
```diff
@@ -31,39 +31,54 @@ package v1alpha1
...
+	// TaskTLP is a top level property supported by CAS template engine
 	//
-	// The task related properties as placed with TaskTLP
-	// as the top level property
-	TaskTLP TopLevelProperty = "Task"
-	// TaskResultTLP is a top level property supported by
-	// volume policy engine
+	// The task related properties as placed with TaskTLP as the top level
+	// property
+	//TaskTLP TopLevelProperty = "Task"
```
Comment (avishnu) [inline pkg/apis/openebs.io/v1alpha1/cas_template_keys.go]: Why is this commented? Is it not required?

File: pkg/task/fetch.go [hunk 7]
```diff
@@ -19,15 +19,15 @@ package task
 import (
 	"fmt"
 
-	"github.com/golang/glog"
+	//"github.com/golang/glog"
```
Comment (avishnu) [inline pkg/task/fetch.go]: Better to remove this line?

File: pkg/task/meta.go [hunk 8]
```diff
@@ -154,7 +166,7 @@ func (m *metaTaskExecutor) getRunNamespace() string {
 func (m *metaTaskExecutor) getRetry() (attempts int, interval time.Duration) {
 	retry := m.metaTask.Retry
 	// "attempts,interval" format
-	defRetry := "1,0s"
+	defRetry := "0,0s"
```
Comment (avishnu) [inline pkg/task/meta.go]: Why this change?

File: pkg/task/runner.go [hunk 9]
```diff
@@ -25,26 +25,32 @@ import (
 	"github.com/openebs/maya/pkg/util"
 )
 
+// redactJsonResult will update the provided map by removing the original json
+// result doc i.e. bytes and replace it with "--redacted--"
+//
+// NOTE:
+//  This should be done once the the task group runner has finished executing all
```
Comment (avishnu) [inline pkg/task/runner.go]: extra "the"

File: pkg/task/task.go [hunk 10]
```diff
@@ -74,104 +83,303 @@ type taskExecutor struct {
...
+			m.resetK8sClient(resource)
+		}
+
+		// set the currently active repeat resource
+		util.SetNestedField(m.templateValues, resource, string(v1alpha1.ListItemsTLP), string(v1alpha1.CurrentRepeatResourceLITP))
+
+		// execute the task function
+		err = m.retryOnVerificationError()
+		if err != nil {
+			// stop repetition on unhandled runtime error & return
```
Comment (avishnu) [inline pkg/task/task.go]: Is this a TODO ?

File: pkg/template/template.go [hunk 11]
```diff
@@ -19,37 +20,621 @@ package template
...
+//  {{- "pkey=openebs--stor2=mstor" | splitList " " | splitKeyMap "-- =" "vals" .Target | noop -}}
+//
+// Above will result into following:
+//  Target: map[string]interface{}{
+//    "vals": map[string]interface{}{
+//      "openebs": map[string]interface{}{
+//        "stor1": "jiva",
+//        "stor2": "cstor, mstor",
+//      },
+//      "pkey": map[string]interface{}{
```
Comment (avishnu) [inline pkg/template/template.go]: Is the example correct? should it be something else and not "pkey" in this line?

File: pkg/template/template.go [hunk 12]
```diff
@@ -19,37 +20,621 @@ package template
...
+		primaryKey string
+		key        string
+		value      string
+		destFields []string
+		fields     []string
+	)
+
+	// defaultPKey is the default primary key if primary key (to build the
+	// maps) is not specified
+	defaultPKey := "pkey"
```
Comment (avishnu) [inline pkg/template/template.go]: Does this address my previous comment?

File: pkg/task/meta.go (hunk 8 above)
Comment (AmitKumarDas) [reply pkg/task/meta.go]: retry & original execution are treated as separate. retry will happen only if original execution has `verify error`. retry will happen if retry value is set.

File: pkg/template/template.go (hunk 1 above)
Comment (AmitKumarDas) [reply pkg/template/template.go]: the purpose is different here. this is a template function that accepts error message in case the verification fails. if there are no error message provided then the default message is "verification failed". this default message is only applicable to this particular function.

File: pkg/template/template.go (hunk 2 above)
Comment (AmitKumarDas) [reply pkg/template/template.go]: in templating it is advised to handle the errors than letting them bubble out as runtime errors. I will try to manage these errors in a way so that it does not bubble out as un-handled error.

File: pkg/template/template.go (hunk 12 above)
Comment (princerachit) [reply pkg/template/template.go]: Yes. If a value to `pkey` is not supplied then the entries would be put in a map whose key is `pkey`

File: pkg/template/template.go (hunk 11 above)
Comment (princerachit) [reply pkg/template/template.go]: Example is correct.

File: pkg/template/template.go (hunk 1 above)
Comment (gkGaneshR) [reply pkg/template/template.go]: ok

File: pkg/task/fetch.go (hunk 7 above)
Comment (AmitKumarDas) [reply pkg/task/fetch.go]: done

File: pkg/template/template.go (hunk 11 above)
Comment (AmitKumarDas) [reply pkg/template/template.go]: this is correct. if there is no primary key in a given string i.e. a k:v pair with k as `pkey` then this function will default the primary key as "pkey"

File: pkg/template/template.go (hunk 12 above)
Comment (AmitKumarDas) [reply pkg/template/template.go]: You are right @avishnu

**Comment by princerachit** [comment]
lgtm

**Comment by avishnu** [comment]
LGTM.
